File: python/sentinel-api/sentinel_api/engram/session.py
"""Builder-pattern session recorder for engram capture."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any

from sentinel_api.engram.models import (
    Action,
    Alternative,
    Decision,
    Engram,
    EngramId,
)

if TYPE_CHECKING:
    from uuid import UUID


class EngramSession:
    """Incrementally records agent reasoning during execution.

    Usage::

        session = EngramSession(tenant_id, "scanner", "Scan subnet")
        session.set_context({"subnet": "10.0.1.0/24"})
        session.add_decision("ICMP scan", "fastest", 0.9)
        session.add_action("scan", "ping sweep", {"hosts": 254}, True)
        engram = session.finalize()
    """

    def __init__(self, tenant_id: UUID, agent_id: str, intent: str) -> None:
        self._engram = Engram(
            tenant_id=tenant_id,
            agent_id=agent_id,
            intent=intent,
        )

    @property
    def id(self) -> EngramId:
        return self._engram.id

    def set_context(self, context: Any) -> None:
        self._engram.context = context

    def add_decision(
        self, choice: str, rationale: str, confidence: float
    ) -> None:
        self._engram.decisions.append(
            Decision(choice=choice, rationale=rationale, confidence=confidence)
        )

    def add_alternative(self, option: str, rejection_reason: str) -> None:
        self._engram.alternatives.append(
            Alternative(option=option, rejection_reason=rejection_reason)
        )

    def add_action(
        self,
        action_type: str,
        description: str,
        details: Any = None,
        success: bool = True,
    ) -> None:
        self._engram.actions.append(
            Action(
                action_type=action_type,
                description=description,
                details=details,
                success=success,
            )
        )

    def finalize(self) -> Engram:
        """Finalize the session: set completed_at and compute content hash."""
        self._engram.completed_at = datetime.now(UTC)
        self._engram.content_hash = self._engram.compute_hash()
        return self._engram

File: crates/sentinel-engram/src/session.rs
//! Builder-pattern session recorder for engram capture.
//!
//! Used by agent code to incrementally record reasoning during execution:
//!
//! ```no_run
//! # use sentinel_engram::session::EngramSession;
//! # use uuid::Uuid;
//! let mut session = EngramSession::new(
//!     Uuid::new_v4(),
//!     "discover-scanner",
//!     "Scan subnet 10.0.1.0/24 for new assets",
//! );
//! session.set_context(serde_json::json!({"subnet": "10.0.1.0/24"}));
//! session.add_decision("Use ICMP + TCP SYN scan", "Fastest for initial discovery", 0.9);
//! session.add_alternative("Full TCP connect scan", "Too slow for /24 subnet");
//! session.add_action("network_scan", "ICMP ping sweep", serde_json::json!({"hosts": 254}), true);
//! let engram = session.finalize();
//! assert!(engram.content_hash.is_some());
//! ```

use chrono::Utc;

use crate::{Action, Alternative, Decision, Engram, EngramId};

/// A session builder that records agent reasoning incrementally.
pub struct EngramSession {
    engram: Engram,
}

impl EngramSession {
    /// Start a new engram recording session.
    pub fn new(tenant_id: uuid::Uuid, agent_id: &str, intent: &str) -> Self {
        Self {
            engram: Engram {
                id: EngramId::new(),
                tenant_id,
                agent_id: agent_id.to_string(),
                intent: intent.to_string(),
                context: serde_json::Value::Null,
                decisions: Vec::new(),
                alternatives: Vec::new(),
                actions: Vec::new(),
                started_at: Utc::now(),
                completed_at: None,
                content_hash: None,
            },
        }
    }

    /// Set the context provided to the agent.
    pub fn set_context(&mut self, context: serde_json::Value) {
        self.engram.context = context;
    }

    /// Record a decision the agent made.
    pub fn add_decision(&mut self, choice: &str, rationale: &str, confidence: f64) {
        self.engram.decisions.push(Decision {
            choice: choice.to_string(),
            rationale: rationale.to_string(),
            confidence,
            timestamp: Utc::now(),
        });
    }

    /// Record an alternative that was considered but not chosen.
    pub fn add_alternative(&mut self, option: &str, rejection_reason: &str) {
        self.engram.alternatives.push(Alternative {
            option: option.to_string(),
            rejection_reason: rejection_reason.to_string(),
        });
    }

    /// Record an action taken by the agent.
    pub fn add_action(
        &mut self,
        action_type: &str,
        description: &str,
        details: serde_json::Value,
        success: bool,
    ) {
        self.engram.actions.push(Action {
            action_type: action_type.to_string(),
            description: description.to_string(),
            details,
            success,
            timestamp: Utc::now(),
        });
    }

    /// The engram ID for this session (available before finalization).
    pub fn id(&self) -> EngramId {
        self.engram.id
    }

    /// Finalize the session: set completed_at and compute the content hash.
    pub fn finalize(mut self) -> Engram {
        self.engram.completed_at = Some(Utc::now());
        let hash = self.engram.compute_hash();
        self.engram.content_hash = Some(hash);
        self.engram
    }
}
